src/Database/Database.php

```diff
@@ -379,8 +386,14 @@ public function query(string $query, array $bindings = [], array $params = [])
 
 		// apply the fetch closure to all results if given
 		if ($options['fetch'] instanceof Closure) {
-			foreach ($results as $key => $result) {
-				$results[$key] = $options['fetch']($result, $key);
+			if ($options['method'] === 'fetchAll') {
+				// fetching multiple records
+				foreach ($results as $key => $result) {
+					$results[$key] = $options['fetch']($result, $key);
+				}
+			} elseif ($options['method'] === 'fetch' && $results !== false) {
+				// fetching a single record
+				$results = $options['fetch']($results);
```

Comment:
Not sure if I would tie this as closely to the exact methods. How about
`is_iterable($results)` for the first if and just `$results !== false` for the elseif?

Reply:
Can't do - the method is the differentiator. the fetch (`->first()` et al) returns:

```
[
	'column1' => 'value',
	'column2' => 'value'
]
```

which is obviously passes `is_iterable()` as well, except now we're trying to send 'value' to closure expecting array.

Reply:
Ah, right, of course. Each row is an array.